src/parallel/ScatterSplitter.cc

```diff
@@ -0,0 +1,369 @@
+            continue;
+        int level = 0;
+        vec<PTRef> clause;
+        bool hasForeignAssumption = false;
+        for (unsigned int j = 0; j < c.size(); j++) {
+            Lit l = c[j];
+            Var v = var(l);
+            if (isAssumptionVar(v)) {
+                vec<opensmt::pair<int, int>> const & solverBranch_perVar = get_solverBranch(v);
+                if (isPrefix(solverBranch_perVar, get_solver_branch())) {
+                    int result = solverBranch_perVar.size();
+                    assert([&]() {
+                    if (result <= 0)
+                    {
+                        std::scoped_lock<std::mutex> s_lk(getChannel().getMutex());
+                        throw PTPLib::common::Exception(__FILE__, __LINE__, ";assert: level is less than zero " +
+                                getChannel().get_current_header().at(PTPLib::common::Param.NODE)+ std::to_string(result));
+                    }
+                    if (result > get_solver_branch().size()) {
+                        std::scoped_lock<std::mutex> s_lk(getChannel().getMutex());
+                        throw PTPLib::common::Exception(__FILE__, __LINE__, ";assert: level is greater than solver address length " +
+                                getChannel().get_current_header().at(PTPLib::common::Param.NODE)+ std::to_string(result));
+                    }
+                        return true;
+                    }());
+
+                    level = std::max<int>(level, result);
+                    continue;
+                } else {
+                    hasForeignAssumption = true;
+                    break;
+                }
+// Second algorithm of how to correctly publish clauses so that they do not include assumption literals.
+// Issue: https://github.com/MasoudAsadzade/OpenSMT2/issues/1
+//                int result = getAssumptionLevel(v);
+//                if (result >= 1) {
+//                    assert(result >= 1);
+//                    level = std::max<int>(level, result);
+//                    continue;
+//                } else {
+//                    hasForeignAssumption = true;
+//                    break;
+//                }
+            }
+            PTRef pt = theory_handler.varToTerm(v);
+            pt = sign(l) ? theory_handler.getLogic().mkNot(pt) : pt;
+            clause.push(pt);
+        }
+        if (hasForeignAssumption or clause.size() > 3)
+            continue;
+        std::string str = this->theory_handler.getLogic().dumpWithLets(theory_handler.getLogic().mkOr(clause));
+
+        learnedLemmas.emplace_back(PTPLib::net::Lemma(str, level));
+        assert([&](std::string_view clause_str) {
+            if (clause_str.find(".frame") != std::string::npos) {
+                throw PTPLib::common::Exception(__FILE__, __LINE__,";assert: frame caught in actual clauses");
+            }
+            return true;
+        }(str));
+        c.mark(3);
```

Comment:
Yes, in here:
https://github.com/MasoudAsadzade/opensmt-verification/blob/429f688c5bb18340ecd147b3af94eba7c71a6e5c/src/parallel/ScatterSplitter.cc#L234
To avoid publishing clauses which already published.

Reply:
Ok!